Lance/topics

LGG233 opened the pull request:
@kylewhitaker -

The functionality for Topics is ready to show you (read: get your feedback so I can improve to the level needed to actually merge this branch into `master`...). 

- Search topics are displayed in My Profile. 
- Users can add and delete topics
- I haven't included a seed file of topics since it should be just as easy to populate your db with real data as with the seed.
- I still haven't mastered the `navigate` function - maybe we can discuss on our next call? I'd like `userProfile` to pull fresh data from the `user` db when it re-renders after user adds a topic to the db (it renders without doing the API call so the new item doesn't appear until a hard refresh is done). Same thing when user deletes a topic from the db. 

A couple of related comments:

- After seeing the way we display data am considering breaking My Profile into three separate components. Should be easy to do but didn't want to include in this PR.
- You'll note that I added an icon to and tidied up the NavBar. What do you think?

kylewhitaker commented on package.json:
```diff
@@ -20,5 +20,8 @@
     "singleQuote": true
+  },
+  "dependencies": {
+    "react-bootstrap": "^1.0.1"
```
There should be zero dependencies in the root package.json. This one should reside in `client/package.json`.

kylewhitaker commented on server/routes/topics.js:
```diff
@@ -0,0 +1,29 @@
+                    res.json(dbTopics);
+                });
+        }
+    });
```
See comment from Pubs PR: https://github.com/LGG233/trending-news/pull/12#discussion_r418085326

if err, then 500
if topic, then 409
otherwise, 200

kylewhitaker commented on server/routes/topics.js:
```diff
@@ -0,0 +1,29 @@
+    });
+});
+
+router.post('/:name', (params, res, next) => {
```
HTTP standards would ask for a DELETE verb since we are deleting a resource. In code, this means a change to use `router.delete` in the server and then also calling `ApiService.delete` from the client.

kylewhitaker commented on client/src/components/topics.jsx:
```diff
@@ -1,38 +1,44 @@
+          <button
+            className="btn btn-link text-secondary"
+            onClick={function () {
+              navigate('./topics-entry');
```
`./` invokes relative routing, which adds complexity. Let's stick with absolute routing by starting every navigation with `/`.

kylewhitaker commented on client/src/components/topics.jsx:
```diff
@@ -1,38 +1,44 @@
 class MyTopics extends Component {
+  deleteTopic(value) {
+    let searchTerm = value.topics;
+    ApiService.post('topics/' + searchTerm)
```
See my other comment re: delete. This should be `ApiService.delete`.

LGG233 replied:
I get an error after I changed to ApiService.delete:

"TypeError: _services__WEBPACK_IMPORTED_MODULE_2__.ApiService.delete is not a function"

kylewhitaker replied:
@LGG233 You have to add the `delete` method to our `ApiService`. We haven't needed it until now.

LGG233 replied:
IT WORKS!!!

kylewhitaker commented on client/src/components/topics.jsx:
```diff
@@ -1,38 +1,44 @@
+              <tr key={index}>
+                <th scope="row">
+                  {topics}
+                  <button className="btn btn-link text-secondary" onClick={() => this.deleteTopic({ topics })}>
```
When using the `.map(...)` function of an Array, I recommend using the singular form of the word to avoid confusion. Here, `topics` should be renamed to `topic` in the callback function. Note that `this.props.topics` does not change...

kylewhitaker commented on client/src/components/pub-display.jsx:
```diff
@@ -67,7 +58,7 @@ class PubDisplay extends Component {
             <tbody>
               {this.state.data.map((pubs) => (
-                <tr>
+                <tr key={pubs._id}>
```
Nice!

kylewhitaker commented on client/src/components/topics-search.jsx:
```diff
@@ -0,0 +1,196 @@
+  }
+
+  async componentDidMount() {
+    await ApiService.get('topics').then(
```
I don't see a corresponding `router.get('/')` in the topics router.

LGG233 replied:
I ended up not using this component. Because there is no longer a master list of topics and search terms (as there is with publications), we don't need to search for a topic. Users can just add all the topics they want, and if the topic is already in their list, they get the 409 error. Do you think I should be more specific on the error, with a separate component that says that "'X' is already in your list of search terms:?

kylewhitaker replied:
For now, handle the same as if it were added. We can enhance later if needed.

kylewhitaker commented on client/src/components/topics-search.jsx:
```diff
@@ -0,0 +1,196 @@
+        <div>
+          <h4>Search Topics</h4>
+          <div className="container-fluid">
+            <DropdownButton id="dropdown-item-button" title={this.state.dropDownValue}>
+              {this.state.data.map((topic, index) => (
+                <Dropdown.Item key={index}>
+                  <div onClick={this.changeValue}>{topic.name}</div>
+                </Dropdown.Item>
+              ))}
+            </DropdownButton>
```
React Bootstrap... here we go!! 🚀

LGG233 replied:
I was pretty pleased with this and a bit disappointed when I got rid of the topic search functionality (that's why I didn't delete the component...). But it quickly became unnecessarily complicated

kylewhitaker commented on server/routes/topics.js:
```diff
@@ -0,0 +1,29 @@
+const User = require('../database/models/user');
+
+router.post('/', (req, res, next) => {
+    User.findOne({ topics: req.body.entryTerm }, (err, topic) => {
```
Order of operations to allow a user to add a new topic to their list:
1. Identify the user by the `username`. Every server request (since Auth v2 changes) now includes an authorization token that is validated & decoded with the `cognito:username`. See the User route for an example of how this works.
1. Check if the new topic already exists for that user. If so, return an error code with `res.sendStatus(409)`. The client could probably gracefully handle this error by doing the same exact thing as if the new topic was added succesfully.
1. If the topic does not exist, add it to the database and return the new list of topics. I'm not sure exactly how Mongo works in this case, but I'm guessing `dbTopics` actually represents a User object here.

LGG233 replied:
My comments / questions (in a somewhat random order):
1. Do I need to identify the user since he/she can't get to this screen without being logged in? The add / delete topics operations take place with the user's Mongo document. If they're not logged in, those operations have no place to go. Not trying to avoid your Step 1, just trying to understand why it's necessary. 
2. Your point about gracefully handling the error makes sense (and resolves my question above): the user doesn't need to know that they forgot that they already put that topic into their database. We have to make sure there are no duplicates, but they don't need to go through an extra step - whether or not the topic exists, we return a `dbTopics` that contains that item. 
3. Yes, `dbTopics` is the object that gets plugged into the profile

kylewhitaker replied:
—1. Let’s discuss over Zoom. Yes, we must identify the user somehow to know which user’s topics to update. Grabbing the username from the authorization token is a secure way to do this.

—3. I ask about the variable name “dbTopics” because if it actually represents a user object, then I’d rename it to “user” to avoid confusion.

LGG233 commented:
OK, let's find a time when we can Zoom - what's your schedule like the rest of this week? BTW, I tried to push my changes to this point and got the following error: 

`To github.com:LGG233/trending-news.git
 ! [rejected]        lance/topics -> lance/topics (non-fast-forward)
error: failed to push some refs to 'git@github.com:LGG233/trending-news.git'`
`hint: Updates were rejected because the tip of your current branch is behind`
`hint: its remote counterpart. Integrate the remote changes (e.g.`
`hint: 'git pull ...') before pushing again.`
`hint: See the 'Note about fast-forwards' in 'git push --help' for details.`

@lgg233 force push the branch, it’s probably because of a rebase

LGG233 commented:
OK, I think this PR is ready to go. Can you give it a final review? Thank you, and thanks so much for your time and help today!

kylewhitaker commented on package.json:
```diff
@@ -21,5 +21,6 @@
   "prettier": {
     "printWidth": 120,
     "singleQuote": true
-  }
+  },
+  "dependencies": {}
 }
```
Undo changes to this file, since there are none, so it doesn't show up as having been modified in this PR

kylewhitaker commented:
@LGG233 Prettier check failed. Click "details" and inspect the logs to see which files failed the check....or just run the command to fix them all and push.

LGG233 commented:
OK, I fixed package.json and committed it along with one other file that had been prettierfied. The command that failed in the check is `npm run formatcheck`. But maybe the Prettier check isn't necessary anymore since we fixed the VSCode setting? 

Either way, I'm going to squash and merge the topics branch so I can start working on getting publications into my user profile. 

BTW, I added a second user and everything works. Thanks again for all your help!

kylewhitaker commented:
[image: image.png]

On Fri, May 22, 2020 at 11:52 PM Lance Godard <notifications@github.com>
wrote:

> OK, I fixed package.json and committed it along with one other file that
> had been prettierfied. The command that failed in the check is npm run
> formatcheck. But maybe the Prettier check isn't necessary anymore since
> we fixed the VSCode setting?
>
> Either way, I'm going to squash and merge the topics branch so I can start
> working on getting publications into my user profile.
>
> BTW, I added a second user and everything works. Thanks again for all your
> help!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LGG233/trending-news/pull/16#issuecomment-632980169>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAZC4LKDMFNRZZFRTVJQMTRS5B7NANCNFSM4M4JR74Q>
> .
>
